src/main/role.ts

```diff
@@ -0,0 +1,46 @@
+export enum RoleStatus {
+    REMOVED,
+    EXISTING,
+    NEW
+}
+
+/**
+ * Gets a list of roles from a roleTodoIdMap
+ * 
+ * @param roleTodoIdMap a map associating an event's roles with todo ids
+ */
+function getOriginalEventRoles(roleTodoIdMap: RoleTodoIdMap): string[] {
+    return Object.keys(roleTodoIdMap);
+}
+
+/**
+ * Gets a list of roles from a roleRequestMap map
+
+ * @param roleRequestMap a map associating role titles with BasecampTodoRequest objects
+ */
+function getCurrentEventRoles(currentRoleRequestMap: RoleRequestMap): string[] {
+    return Object.keys(currentRoleRequestMap);
+}
+
+/**
+ * Gets a list of roles (i.e. obsolete, new, surviving) by comparing the original and current roles together
+ * 
+ * @param roleRequestMap a map associating an event's roles with api request bodies
+ * @param currentRoleTodoIdMap a map associating an event's roles to existing todo ids, that is currently saved in the document properties
+ * @returns an array of obsolete roles
+ */
+export function getRoles(roleRequestMap: RoleRequestMap, currentRoleTodoIdMap: RoleTodoIdMap, roleStatus: RoleStatus): string[] {
+    const currentRoles: string[] = getCurrentEventRoles(roleRequestMap);
+    const originalRoles: string[] = getOriginalEventRoles(currentRoleTodoIdMap);
+
+    switch(roleStatus) {
```

Comment:
best practice is to avoid "flag" arguments for methods. this is because it unnecessarily complicates a function (even though in this example it's not that complex), since the alternative is to have 3 different simple functions, appropriately named (getRemovedRoles, getExistingRoles, getNewRoles) which is far more clear.
and best practice is also to minimize the number of arguments in a function as much as possible, as it makes testing more work.

Reply:
I see your point how avoiding flag arguments for methods and breaking out this function into 3 separate methods makes things clearer. However, the reason why I merged them in the first place is that I didn't like how I would have to repeat the getCurrentEventRoles() and getOriginalEventRoles() methods over and over again:

`/**
 * Gets a list of obsolete roles by comparing the original and current roles
 * 
 * @param roleRequestMap a map associating an event's roles with API request bodies
 * @param lastSavedRoleTodoIdMap a map associating an event's roles to existing todo IDs, that is currently saved in the document properties
 * @returns an array of removed roles
 */
export function getRemovedRoles(roleRequestMap: RoleRequestMap, lastSavedRoleTodoIdMap: RoleTodoIdMap): string[] {
    const currentRoles: string[] = getCurrentEventRoles(roleRequestMap);
    const originalRoles: string[] = getOriginalEventRoles(lastSavedRoleTodoIdMap);
    return originalRoles.filter(role => !currentRoles.includes(role));
}

/**
 * Gets a list of existing roles by comparing the original and current roles
 * 
 * @param roleRequestMap a map associating an event's roles with API request bodies
 * @param lastSavedRoleTodoIdMap a map associating an event's roles to existing todo IDs, that is currently saved in the document properties
 * @returns an array of existing roles
 */
export function getExistingRoles(roleRequestMap: RoleRequestMap, lastSavedRoleTodoIdMap: RoleTodoIdMap): string[] {
    const currentRoles: string[] = getCurrentEventRoles(roleRequestMap);
    const originalRoles: string[] = getOriginalEventRoles(lastSavedRoleTodoIdMap);
    return currentRoles.filter(role => originalRoles.includes(role));
}

/**
 * Gets a list of new roles by comparing the original and current roles
 * 
 * @param roleRequestMap a map associating an event's roles with API request bodies
 * @param lastSavedRoleTodoIdMap a map associating an event's roles to existing todo IDs, that is currently saved in the document properties
 * @returns an array of new roles
 */
export function getNewRoles(roleRequestMap: RoleRequestMap, lastSavedRoleTodoIdMap: RoleTodoIdMap): string[] {
    const currentRoles: string[] = getCurrentEventRoles(roleRequestMap);
    const originalRoles: string[] = getOriginalEventRoles(lastSavedRoleTodoIdMap);
    return currentRoles.filter(role => !originalRoles.includes(role));
}`

Is there a way to get around that?

Reply:
I can still break out the getRoles method into the three functions, but I'm just wondering if there's a way to address the repeated code? I could try calling these methods two levels up and pass the currentRoles and originalRoles arrays into the calling functions for getNewRoles, getExistingRoles, and getRemovedRoles however that would increase the argument counts for each of the calling functions to 4, which I think is a lot

Reply:
I broke out getRoles() into the three methods that you suggested

Reply:
you're right that there's repeated code. and that is a good principle too (Don't Repeat Yourself), but in this case it's only 2 lines of code repeated, which is pretty insignificant. my opinion here is that the clarity that comes from removing the flag argument outweighs the 2 lines of repeated code.
i also thought about the possibility of calling those functions earlier. you're also right that passing in 4 things is a lot. the general way to get around this is to call it up earlier and wrap all 4 things necessary into its own object, and pass that. in this case, it might be a bit too over-engineered (unless this type of object really is used a lot), and to solve for just 2 lines of repeated code, it's probably not necessary